Replacing AES impls with interop

nmahendru opened the pull request:
#### 7f4c5caad9baf65e6b1192b1f3826519980f680d
<pre>
Replacing AES impls with interop
<a href="https://bugs.webkit.org/show_bug.cgi?id=272472">https://bugs.webkit.org/show_bug.cgi?id=272472</a>
<a href="https://rdar.apple.com/126217685">rdar://126217685</a>

Reviewed by Elliott Williams.

Revert &quot;Unreviewed, reverting 277285@main&quot;

This reverts commit 9f89447abd9a1625c7a2e4f04ef7cb04a64d8c0d.
It also adds build flags to elevate the build targets for interop.
It also removes the ASSERTION for Vector Copy constructor until we find a better
solution.

Old Commit:

All the Unsafe Bits are now in UnsafeOverlays.swift.
Also all the Cpp wrappers to work around interop shortcomings are now
inside a namespace to keep them away from global scope.

Any wrappers deemed Unsafe are now inside namespace Cpp.Unsafe.

WTF::Vector Copy operation is hard fail from Swift. This is to guard
against swift&apos;s strategy of implicity copies of C++ types for exclusivity and
memory safety.

Also the quirk added in 276651@main for destructor linkage
 is moved out of the execution path.

* Source/WTF/wtf/Vector.h:
(WTF::Malloc&gt;::Vector):
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/pal/PALSwift.h:
(Cpp::toOptionalVectorUInt8):
(Cpp::createVectorUInt8FromSpan):
(Cpp::appendSpanToVectorUInt8):
(Cpp::Unsafe::createConstUInt8Span):
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(AesGcmRV.cipherText):
(AesGcmRV.errorCode):
(AesGcm.ciphertext):
(AesKwRV.result):
(AesKw.wrap(_:keyToBeWrapped:)):
(AesKw.unwrap(_:wrappedKey:)):
(AesKwRV.outputSize): Deleted.
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
(HashFunction.update(_:)):
(ContiguousBytes.createConstUInt8Span(_:)):
(ContiguousBytes.createConstUInt8Span):
(VectorUInt8.appendSpan(_:)):
(VectorUInt8.appendContiguousBytes(_:)):
(Data.fromSpan(_:)):
(_CompilerBugDontUse.forceLinkageForVectorDestructor):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
(WebCore::encryptCryptoKitAESGCM):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp:
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):

Canonical link: <a href="https://commits.webkit.org/277329@main">https://commits.webkit.org/277329@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ff20896f0d79879331873d382e28b00d4fc21b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47341 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49992 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50025 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43390 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49648 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32033 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23981 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38546 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47922 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24057 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19866 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21493 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41962 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5385 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43691 "Passed tests") | [⏳ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51900 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22371 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18712 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45845 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23646 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44871 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10433 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24431 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23364 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -35,7 +35,8 @@ WK_APPLEJPEGXL_PREPROCESSOR_DEFINITIONS_YES = USE_JPEGXL=1;
 WK_HAVE_SWIFT_CPP_INTEROP = HAVE_SWIFT_CPP_INTEROP=0;
-WK_HAVE_SWIFT_CPP_INTEROP[sdk=macos14*] = ;
+WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14.4*] = ;
+WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14.5*] = ;
```
This is unfortunate, because it means we're going to have to update the config for every new `14.x` minor release, otherwise interop will silently be disabled. How about this:
```suggestion
WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14*] = ;
WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14.0*] = HAVE_SWIFT_CPP_INTEROP=0;
WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14.1*] = HAVE_SWIFT_CPP_INTEROP=0;
WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14.2*] = HAVE_SWIFT_CPP_INTEROP=0;
WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14.3*] = HAVE_SWIFT_CPP_INTEROP=0;
```

emw-apple commented on Source/WebCore/PAL/Configurations/PAL.xcconfig:
```diff
@@ -26,7 +26,8 @@
 WK_HAVE_SWIFT_CPP_INTEROP = HAVE_SWIFT_CPP_INTEROP=0;
-WK_HAVE_SWIFT_CPP_INTEROP[sdk=macos14*] = ;
+WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14.4*] = ;
+WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14.5*] = ;
```
Ditto.

emw-apple commented on Source/WebCore/PAL/Configurations/PAL.xcconfig:
```diff
@@ -101,7 +102,8 @@ OTHER_SWIFT_FLAGS_TVOS_SINCE_17 = -Xcc -std=c++20 -I $(SRCROOT)/pal -I $(SRCROOT
 EXCLUDED_SOURCE_FILE_NAMES = CryptoKitShim.swift UnsafeOverlays.swift
-EXCLUDED_SOURCE_FILE_NAMES[sdk=macos14*] = ;
+EXCLUDED_SOURCE_FILE_NAMES[sdk=macosx14.4*] = ;
+EXCLUDED_SOURCE_FILE_NAMES[sdk=macosx14.5*] = ;
```
Ditto.

emw-apple commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1026,6 +1026,9 @@ template<typename T, size_t inlineCapacity, typename OverflowHandler, size_t min
 {
+#ifdef __swift__
+    RELEASE_ASSERT_NOT_REACHED_WITH_MESSAGE("C++ copy constructors are not memory safe in Swift. Try to find a way to borrow the object instead.");
+#endif
```
I think that committing this is a risk per our discussion in https://github.com/WebKit/WebKit/pull/26438#discussion_r1557971121. Seems like it is an ODR violation, and in practice the linker could choose the swift-compiled instantiation arbitrarily and break callers from C++.

nmahendru committed:
Replacing AES impls with interop
https://bugs.webkit.org/show_bug.cgi?id=272472
rdar://126217685

Reviewed by Elliott Williams.

Revert "Unreviewed, reverting 277285@main"

This reverts commit 9f89447abd9a1625c7a2e4f04ef7cb04a64d8c0d.
It also adds build flags to elevate the build targets for interop.
It also removes the ASSERTION for Vector Copy constructor until we find a better
solution.

Old Commit:

All the Unsafe Bits are now in UnsafeOverlays.swift.
Also all the Cpp wrappers to work around interop shortcomings are now
inside a namespace to keep them away from global scope.

Any wrappers deemed Unsafe are now inside namespace Cpp.Unsafe.

WTF::Vector Copy operation is hard fail from Swift. This is to guard
against swift's strategy of implicity copies of C++ types for exclusivity and
memory safety.

Also the quirk added in 276651@main for destructor linkage
 is moved out of the execution path.

* Source/WTF/wtf/Vector.h:
(WTF::Malloc>::Vector):
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/pal/PALSwift.h:
(Cpp::toOptionalVectorUInt8):
(Cpp::createVectorUInt8FromSpan):
(Cpp::appendSpanToVectorUInt8):
(Cpp::Unsafe::createConstUInt8Span):
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(AesGcmRV.cipherText):
(AesGcmRV.errorCode):
(AesGcm.ciphertext):
(AesKwRV.result):
(AesKw.wrap(_:keyToBeWrapped:)):
(AesKw.unwrap(_:wrappedKey:)):
(AesKwRV.outputSize): Deleted.
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
(HashFunction.update(_:)):
(ContiguousBytes.createConstUInt8Span(_:)):
(ContiguousBytes.createConstUInt8Span):
(VectorUInt8.appendSpan(_:)):
(VectorUInt8.appendContiguousBytes(_:)):
(Data.fromSpan(_:)):
(_CompilerBugDontUse.forceLinkageForVectorDestructor):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
(WebCore::encryptCryptoKitAESGCM):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp:
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):

Canonical link: https://commits.webkit.org/277329@main
